### ISSUE TITLE
Bad index on text collection type

### ISSUE THREAD

**Issue by Soullivaneuh** [sonata-project/SonataAdminBundle#2857]
Giving the following Doctrine entity attribute:

```php
/**
 * @var string[]
 *
 * @ORM\Column(type="array", name="press_images_blacklist", nullable=true)
 */
private $pressImagesBlacklist;
```

And the following admin form mapping of this attribute:

```php
$formMapper
    ->add('pressImagesBlacklist', 'collection', [
        'type'          => 'text',
        'allow_add'     => true,
        'allow_delete'  => true,
    ])
;
```

**Step 1:** load an entity for edition with the some strings already on array (toto, tata, titi), you will get this:

![sonata_collection_step_1](https://cloud.githubusercontent.com/assets/1698357/6857984/0fc78a92-d40b-11e4-897a-5de30a646f90.png)

With the following rendered text `input` tag:

```html
<!-- [...] -->
<input type="text" id="s55147f49f21b1_pressImagesBlacklist_1" name="s55147f49f21b1[pressImagesBlacklist][1]" required="required" class=" form-control" value="toto">
<!-- [...] -->
<input type="text" id="s55147f49f21b1_pressImagesBlacklist_2" name="s55147f49f21b1[pressImagesBlacklist][2]" required="required" class=" form-control" value="tata">
<!-- [...] -->
<input type="text" id="s55147f49f21b1_pressImagesBlacklist_3" name="s55147f49f21b1[pressImagesBlacklist][3]" required="required" class=" form-control" value="titi">
<!-- [...] -->
```

**Step 2:** Add a field by clicking on `+` icon and fill it with whatever. You will get this:

![sonata_collection_step_2](https://cloud.githubusercontent.com/assets/1698357/6858002/5a8681dc-d40b-11e4-9dbc-f856d319a701.png)

With the following new `input` tag:

```html
<input type="text" id="s55147f49f21b1_pressImagesBlacklist_1" name="s55147f49f21b1[pressImagesBlacklist][1]" required="required" class=" form-control">
```

Actually you should already see the problem... ;)

**Step 3:** Submit your wonderful form and you will get... this:

![sonata_collection_step_3_result](https://cloud.githubusercontent.com/assets/1698357/6858029/8f2ad38e-d40b-11e4-9835-7f4f295385b0.png)

Interesting isn't it? Very annoying too. :)

The problem is when you want to add a input from an existing list, sonata doesn't care of that and start with a new index, beginning by `1`. Obviously, if you submit that, the first element is overridden by the new element instead of being just added.

My current symfony and sonata installation:

```
symfony/symfony                          2.7.x-dev 350f30b  The Symfony PHP framework
sonata-project/admin-bundle              dev-master 8594eb4 Symfony SonataAdminBundle
sonata-project/block-bundle              2.3.0              Symfony SonataBlockBundle
sonata-project/cache                     1.0.4              Cache library
sonata-project/core-bundle               2.3.1              Symfony SonataCoreBundle
sonata-project/doctrine-orm-admin-bundle 2.3.1              Symfony Sonata / Integrate Doctrine ORM into the SonataAdminBundle
sonata-project/easy-extends-bundle       2.1.9              Symfony SonataEasyExtendsBundle
sonata-project/exporter                  1.3.4              Lightweight Exporter library
sonata-project/intl-bundle               2.2.1              Symfony SonataIntlBundle
```

I'm working on a fix right now. I will submit a PR if I find out and fix the issue.

Thanks.

**Comment by Soullivaneuh** [comment]
Maybe related to #835 but not sure. Didn't understand all of the asking here.

**Comment by EmmanuelVella** [comment]
I think it's related to https://github.com/sonata-project/SonataAdminBundle/pull/2831

**Comment by Soullivaneuh** [comment]
@EmmanuelVella this PR is supposed to fix this pointed issue or you think this issue is caused especially by this PR?

**Comment by Soullivaneuh** [comment]
Apparently it's a fix. Very weird, I'm up to date with master.

**Comment by EmmanuelVella** [comment]
It fixes another issue, but i think it introduces a new one. I think the counter is not initialized correctly when there already are items in the collection.

**Comment by Soullivaneuh** [comment]
@EmmanuelVella Ok will take a look. What the another issue? I don't want to reintroduce it! :)

**Comment by EmmanuelVella** [comment]
Its explained hère :https://github.com/sonata-project/SonataAdminBundle/pull/2831#issuecomment-82862241

**Comment by Soullivaneuh** [comment]
Fixed on #2858.
